Add wear depth to weighted gap for contact dynamics

recuero opened the pull request:
## Reason
- Simulation aborts due to an exception/error in libMesh when NaNs or meaningless numbers are generated.
- Need a simple way to account for wear for small depths (to model, e.g., fretting wear).
- Friction coefficient may take complex values dependent on contact pressure, velocities, etc.

## Design
- Catch `libMesh::LogicError` exception in mortar aux kernel thread.
- Modify weighted gap to include the effect of a relatively small worn depth.
- Use recent development (AD Function) to introduce a variable-dependent friction coefficient in `ComputeDynamicFrictionalForceLMMechanicalContact`.

## Impact
- Allow many types of simulations using mortar auxiliary kernels to complete.
- Account for wear in dynamic simulations
- Enables more accurate modeling of friction

(#19671)

recuero committed:
WIP wear

recuero commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5508,18 +5508,30 @@ FEProblemBase::computeResidualTags(const std::set<TagID> & tags)
   try
   {
-    _aux->compute(EXEC_LINEAR);
+    try
```
@lindsayad Any suggestion? The code here appears to properly catch the exception, but only in serial. In parallel, the simulation just hangs. I thought the `PARALLEL_CATCH` macros you put together around `ComputeMortarNodalAuxBndThread<AuxKernel>` would take care of that.

dschwen replied:
PARALLEL_CATCH only performs a mpi communication of `_has_exception` and `_exception _message` and the rethrows on every processor. For that to work there has t be a try/catch before the parallel catch that calls setExceptionMessage (or whatever it's called)

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/959806/) on 93eb1fd wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/20332/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/959816/) on 93eb1fd wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>fbe1bb</th>
      <th colspan=4>#20332 93eb1f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>82.33%</td>
      <td>82.32%</td>
      <td>-0.01%</td>
      <td>23.53%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>73889</td>
      <td>73738</td>
      <td>-151</td>
      <td>4</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15863</td>
      <td>15838</td>
      <td>-25</td>
      <td>13</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20332/coverage_diff/framework.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20332/coverage/framework)


## Modules coverage

### Functional expansion tools

<table>
  <thead>
    <tr>
      <th></th>
      <th>fbe1bb</th>
      <th colspan=4>#20332 93eb1f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.12%</td>
      <td>83.09%</td>
      <td>-0.03%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>926</td>
      <td>924</td>
      <td>-2</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>188</td>
      <td>188</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20332/coverage_diff/functional_expansion_tools.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20332/coverage/functional_expansion_tools)

### Phase field

<table>
  <thead>
    <tr>
      <th></th>
      <th>fbe1bb</th>
      <th colspan=4>#20332 93eb1f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>90.20%</td>
      <td>90.21%</td>
      <td>+0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>11976</td>
      <td>11972</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1301</td>
      <td>1299</td>
      <td>-2</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20332/coverage_diff/phase_field.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20332/coverage/phase_field)

### Tensor mechanics

<table>
  <thead>
    <tr>
      <th></th>
      <th>fbe1bb</th>
      <th colspan=4>#20332 93eb1f</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.43%</td>
      <td>89.43%</td>
      <td>-0.00%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>26161</td>
      <td>26160</td>
      <td>-1</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>3092</td>
      <td>3092</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/20332/coverage_diff/tensor_mechanics.html)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/20332/coverage/tensor_mechanics)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/contact)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/20332/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 23.53% is less than the suggested 90.0%



This comment will be updated on new commits.

dschwen commented:
Following up on our conversation, @recuero . This is bad:

```
void function()
{
  PARALLEL_TRY
  {
    try {
      // do some stuff that could fail on one or more MPI ranks
    } catch(MooseException & e) {
      // set the exception flag and message for the current rank
      setException(e.what());
    }
  }
  PARALLEL_CATCH
}

void caller()
{
  PARALLEL_TRY
  {
    // call the function above
    function();
  }
  PARALLEL_CATCH
}
```

This can be made safe by always using a `try {} catch(MooseException & e) { setException(e.what()) ;}` inside the `PARALLEL_TRY` block. Otherwise the the second `PARALLEL_CATCH` sees that the `_has_exception` flag is false (it was reset by the first `PARALLEL_CATCH`) and it will just drop the exception rather than rethrowing it.

recuero committed:
Catch libMesh::LogicError in auxiliary system.

This is necessary to avoid simulation aborts as the Cholesky decomposition is invoked
in libMesh to compute dual bases when using mortar auxiliary kernels.

(#19671)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/962719/) on 83fc06b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20332/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20332/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 271672774901a3a7206be1b58dab1d7201402a9d`

dschwen commented on framework/src/systems/AuxiliarySystem.C:
```diff
@@ -736,12 +736,20 @@ AuxiliarySystem::computeMortarNodalVars(const ExecFlagType type)
+            _fe_problem.setException(e.what());
+          }
         }
         PARALLEL_CATCH;
+        solution().close();
+        _sys.update();
```
`solution().close();` will perform MPI communication, however it will never be reached if a libMesh exception is thrown from within `ComputeMortarNodalAuxBndThread`

dschwen commented on framework/src/systems/AuxiliarySystem.C:
```diff
@@ -736,12 +736,20 @@ AuxiliarySystem::computeMortarNodalVars(const ExecFlagType type)
+                _fe_problem, mortar_nodal, bnd_id, index);
+            Threads::parallel_reduce(bnd_nodes, mnabt);
+          }
+          catch (MooseException & e)
+          {
+            _fe_problem.setException(e.what());
+          }
```
ComputeMortarNodalAuxBndThread
```suggestion
          catch (MooseException & e)
          {
            _fe_problem.setException(e.what());
          }
          catch (libMesh::LogicError & e)
          {
            _fe_problem.setException("We caught a libMesh::LogicError.");
          }
```

recuero replied:
Yeah, we can catch the `libMesh::LogicError` in `ComputeMortarNodalAuxBndThread` and do nothing in `FEProblemBase` to avoid parallel communication issues. That seems to work.

dschwen commented on framework/src/problems/FEProblemBase.C:
```diff
@@ -5653,7 +5665,34 @@ FEProblemBase::computeJacobianTags(const std::set<TagID> & tags)
-    _aux->compute(EXEC_NONLINEAR);
+    try
+    {
+      try
+      {
+        _aux->compute(EXEC_NONLINEAR);
+      }
+      catch (libMesh::LogicError & e)
+      {
+        throw MooseException("We caught a libMesh::LogicError in FEProblemBase");
+      }
+    }
```
```suggestion
      _aux->compute(EXEC_NONLINEAR);
    }
```

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/962882/) on ceac915 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20332/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20332/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 271672774901a3a7206be1b58dab1d7201402a9d`

recuero commented:
@dschwen Thanks! Catching the `libMesh::LogicError` in the `AuxiliarySystem` seems to do the trick. Removed all the changes in `FEProblemBase`.

recuero committed:
Add influence of wear in the computation of dynamic contact constraints.

(#19671)

recuero committed:
Add AD function-based friction coefficient for dynamics.

(#19671)

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/963350/) on 7148098 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20332/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20332/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 271672774901a3a7206be1b58dab1d7201402a9d`

recuero committed:
Avoid valgrind contact error (#19671)

I speculate the error is due to running the mortar auxiliary kernel
at executione stages such as INITIAL or LINEAR.

lindsayad commented:
@reucero please request my review when you're ready for it. I'm going to unsubscribe for now

recuero committed:
Fix MortarFrictionalPressureVectorAux for edge cases.

Particularly, when the lower-d mesh does not extend to the entire outer
topology, we may locate an internal element from a node on the secondary surface
that does not have a corresponding lower-d secondary element since it happens to be
outside the contact sideset. Since we were not checking for this, a bad access can
happen, and it does.

Now, we are sure we are obtaining one of the possible lower-d secondary elements.
Still, it doesn't matter which one we are getting since we loop over nodes, so
 any lower-d element does the trick.

lindsayad commented on modules/contact/src/auxkernels/MortarFrictionalPressureVectorAux.C:
```diff
@@ -95,15 +95,10 @@ Real
+  // However, we are looping over nodes below, so we will locate the correct geometry
   const Elem * lower_dimensional_element =
-      _mortar_generation_object->getSecondaryLowerdElemFromSecondaryElem(element_ids[0]);
+      _mortar_generation_object->nodesToSecondaryElem().at(_current_node->id())[0];
```
```suggestion
      libmesh_map_find(_mortar_generation_object->nodesToSecondaryElem(), _current_node->id())[0];
```

lindsayad commented on framework/include/constraints/AutomaticMortarGeneration.h:
```diff
@@ -326,6 +326,14 @@ class AutomaticMortarGeneration : public ConsoleStreamInterface
+  /**
+   * @return Map from node id to secondary lower-d element pointer
+   */
+  const std::unordered_map<dof_id_type, std::vector<const Elem *>> nodesToSecondaryElem() const
```
```suggestion
  const std::unordered_map<dof_id_type, std::vector<const Elem *>> & nodesToSecondaryElem() const
```
Don't want to make a copy here

lindsayad commented on modules/contact/include/constraints/ComputeDynamicFrictionalForceLMMechanicalContact.h:
```diff
@@ -105,6 +113,12 @@ class ComputeDynamicFrictionalForceLMMechanicalContact
   const Real _mu;
 
+  /// input function
+  const Function * _function_friction;
```
```suggestion
  const Function * const _function_friction;
```

lindsayad commented on modules/contact/src/constraints/ComputeDynamicFrictionalForceLMMechanicalContact.C:
```diff
@@ -59,6 +66,16 @@ ComputeDynamicFrictionalForceLMMechanicalContact::ComputeDynamicFrictionalForceL
       "in order to make its implementation feasible");
 #endif
 
+  if (!_has_friction_function && !isParamValid("mu"))
+    paramError(
+        "mu",
+        "A coefficient of friction needs to be provided as a constant value of via a function.");
```
My understanding of `paramError` is that it will give the line number corresponding to the problematic parameter. However, if the parameter doesn't actually exist, I don't know what it does. I don't know if it will error itself, emit unknown and undefined dragons, or work very peacefully.

recuero replied:
It seems to behave like a regular error in that case. Switching to `mooseError`.

lindsayad commented on modules/contact/src/constraints/ComputeDynamicFrictionalForceLMMechanicalContact.C:
```diff
@@ -277,6 +294,10 @@ ComputeDynamicFrictionalForceLMMechanicalContact::enforceConstraintOnDof3d(
   const Real c = _normalize_c ? _c / *_normalization_ptr : _c;
   const Real c_t = _normalize_c ? _c_t / *_normalization_ptr : _c_t;
 
+  // Compute the friction coefficient (constant or function)
+  ADReal mu_ad = 0.0;
+  computeFrictionValue(mu_ad, contact_pressure, *tangential_vel[0], *tangential_vel[1]);
```
This is fine, but could just have `computeFrictionValue` return an `ADReal` instead of being `void`

lindsayad commented on modules/contact/src/constraints/ComputeDynamicFrictionalForceLMMechanicalContact.C:
```diff
@@ -368,3 +393,23 @@ ComputeDynamicFrictionalForceLMMechanicalContact::enforceConstraintOnDof(
+    ADReal tangential_vel_magnitude = std::sqrt(tangential_vel * tangential_vel +
+                                                tangential_vel_dir * tangential_vel_dir + 1.0e-24);
+
+    mu_ad = _function_friction->value<ADReal>(0.0, contact_pressure, tangential_vel_magnitude, 0.0);
```
Leveraging @dschwen 's new work? Neato!

lindsayad commented on modules/contact/src/constraints/ComputeDynamicWeightedGapLMMechanicalContact.C:
```diff
@@ -23,11 +23,16 @@ ComputeDynamicWeightedGapLMMechanicalContact::validParams()
   InputParameters params = ComputeWeightedGapLMMechanicalContact::validParams();
   params.addClassDescription(
       "Computes the normal contact mortar constraints for dynamic simulations");
+  params.addCoupledVar("wear_depth", "The name of the mortar auxiliary variable that ");
```
Incomplete doc string here

lindsayad replied:
```suggestion
```
Oh wait I see it gets added again below

lindsayad replied:
I think we can delete this line

lindsayad commented on modules/contact/src/constraints/ComputeDynamicWeightedGapLMMechanicalContact.C:
```diff
@@ -137,11 +152,16 @@ ComputeDynamicWeightedGapLMMechanicalContact::residualSetup()
 void
 ComputeDynamicWeightedGapLMMechanicalContact::post()
 {
+  // No need to communicate wear; they should be already ghosted (not computed in the constraint)
```
I don't understand this. It seems like you will be missing contributions to nodal wear depths from attached elements that are on other processes

lindsayad replied:
You are computing `_dof_to_nodal_wear_depth` at line 127 in `computeQpIProperties` which is going to be element-based so your computation of `_dof_to_nodal_wear_depth` will be incomplete if not all of the mortar segment elements that contribute to `dof` are on the same process

recuero replied:
You are right. I tried to approach this in a couple of ways and ended up building the `_dof_to_nodal_wear_depth` map.

recuero committed:
Address Alex's review. (#19671)

Communicate wear depth among processors. We have to do this because
we are taking the wear depth computed from the mortar auxiliary kernel and then
integrate it over the mortar segments, which may reside in different processors.
This strategy may not be strictly necessary but it fits right in with the existing
design of mortar contact mechanics constraint classes.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/974630/) on 6b649d6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/20332/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/20332/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5363455ac17e5b47a9e21ec3d5476a97cce87415`

recuero committed:
Fix bad merge.

Also update test. Auxiliary kernel will need at least to be run in linear and
nonlinear iterations. Hopefully this won't break valgrind testing.

lindsayad commented on modules/contact/src/constraints/ComputeDynamicWeightedGapLMMechanicalContact.C:
```diff
@@ -206,3 +241,42 @@ ComputeDynamicWeightedGapLMMechanicalContact::incorrectEdgeDroppingPost(
 }
+
+void
+ComputeDynamicWeightedGapLMMechanicalContact::communicateWear()
```
at some point we might think about consolidating all these communication methods into a single templated method that can be called for all the use cases

recuero replied:
Yeah, that sounds good. It's been used already for 3 different quantities. Will keep it in mind for next time.
